src/pygeoprocessing/geoprocessing.py

```diff
@@ -632,6 +648,120 @@ def array_equals_nodata(array, nodata):
+    Raster inputs are split into aligned blocks, and the function is
+    applied individually to each stack of blocks (as numpy arrays).
+
+    Args:
+        op (function): Function to apply to the inputs. It should accept a
+            number of arguments equal to the length of ``*inputs``. It should
+           return a numpy array with the same shape as its array input(s).
+        *rasters (strs): Paths to rasters to input to ``op``, in the order that
+           they will be passed to ``op``. All rasters should be aligned and
+           have the same dimensions.
+        target_path (str): path to write out the output raster.
+        target_nodata (number): Nodata value to use for the output raster.
+            Optional. If not provided, a suitable nodata value will be chosen.
+        target_dtype (numpy.dtype): dtype to use for the output. Optional. If
+            not provided, a suitable dtype will be chosen.
+
+    Returns:
+        ``None``
+    """
+    nodatas = [get_raster_info(r)['nodata'][0] for r in rasters]
+
+    def apply_op(*arrays):
+        """Apply the function `op` to the input arrays.
+
+        Args:
+            *arrays: numpy arrays with the same shape.
+
+        Returns:
+            numpy array
+        """
+        result = numpy.full(arrays[0].shape, target_nodata, dtype=target_dtype)
+
+        # make a mask that is True where all input arrays are valid,
+        # and False where any input array is invalid.
+        valid_mask = numpy.full(arrays[0].shape, True)
+        for array, nodata in zip(arrays, nodatas):
+            valid_mask &= ~array_equals_nodata(array, nodata)
+
+        # mask all arrays to the area where they all are valid
+        masked_arrays = [array[valid_mask] for array in arrays]
+        # apply op to the masked arrays in order
+        result[valid_mask] = op(*masked_arrays)
+        return result
+
+    # choose an appropriate dtype if none was given
+    if target_dtype is None:
+        target_dtype = choose_dtype(*rasters)
+
+    # choose an appropriate nodata value if none was given
+    # if the user provides a target nodata,
+    # check that it can fit in the target dtype
+    if target_nodata is None:
+        target_nodata = choose_nodata(target_dtype)
+    else:
+        if not numpy.can_cast(target_nodata, target_dtype):
+            raise ValueError(
+                f'Target nodata value {target_nodata} is incompatible with '
+                f'the target dtype {target_dtype}')
+
+    creation_options = DEFAULT_CREATION_OPTIONS
+    if target_dtype == numpy.int8:
+        creation_options = INT8_CREATION_OPTIONS
+
+    raster_calculator(
+        [(path, 1) for path in rasters],  # assume the first band
+        apply_op,
+        target_path,
+        NUMPY_TO_GDAL_TYPE[numpy.dtype(target_dtype)],
+        target_nodata,
+        raster_driver_creation_tuple=('GTIFF', tuple(creation_options)))
```

Comment:
While creating a GeoTIFF is probably the most common use case, it is a little inconsistent with the existing raster_calculator function (and some other functions).  Would you be opposed to exposing an option for raster driver creation options, but defaulting to GTiff?  Of course we'd still need to handle the signed byte case for GTiffs as well.

Reply:
Oh interesting, I agree that other formats should be allowed. In `warp_raster`, it looks like only the first element (the format) of `raster_driver_creation_tuple` is used - the other creation options are defined with a default like this.

Reply:
I added a `target_driver` kwarg to `raster_map` to expose this option. The benefit of _not_ exposing the other creation options is that `raster_map` can handle setting the int8 `SIGNED_BYTE` flag if needed. Do you think that's sufficient for the `raster_map` use case, or should we expose the creation options tuple as well?

Reply:
Yeah, I think we need to expose the `creation_options` tuple as well.  As implemented, the defaults look like this in `geoprocessing_core.pyx`:

```python
DEFAULT_CREATION_OPTIONS = ('TILED=YES', 'BIGTIFF=YES', 'COMPRESS=LZW',
                             'BLOCKXSIZE=256', 'BLOCKYSIZE=256')
INT8_CREATION_OPTIONS = DEFAULT_CREATION_OPTIONS + (
    'PIXELTYPE=SIGNEDBYTE',)
```

The likely conflict here is that `DEFAULT_CREATION_OPTIONS` are GTiff-specific (here's the [ERS format](https://gdal.org/drivers/raster/ers.html), for example, with many fewer creation options).

`PIXELTYPE=SIGNEDBYTE` is a GDAL implementation detail that happens to be consistent across the data formats that support signed bytes, which are [EEDAI, ERS, GTA, GTiff, HFA, netCDF, PostGISRaster, EHDR, RRaster, Rasterlite2 and Zarr.](https://gdal.org/development/rfc/rfc87_signed_int8.html#details).

So, it makes sense that `PIXELTYPE=SIGNEDBYTE` would be added to the needed creation options in certain cases, but the base set of creation options should be specific to the target datatype, and since the user can change the datatype they should also be able to change the creation options.

Reply:
Oh, thanks for pointing that out. I added a `raster_driver_creation_tuple` argument consistent with other pygeoprocessing functions, but `PIXELTYPE=SIGNEDBYTE` will be added if necessary.